kf_task_fhir_etl/etl/ingest.py

```diff
@@ -0,0 +1,525 @@
+                study_all_targets.add(VitalStatus)
+
+            # biospecimen-diagnoses
+            biospecimen_diagnoses = study_mapped_df_dict.get("biospecimen-diagnoses")
+            if biospecimen_diagnoses is not None:
+                biospecimen_diagnoses = biospecimen_diagnoses.rename(
+                    columns={
+                        "biospecimen_id": CONCEPT.BIOSPECIMEN.TARGET_SERVICE_ID,
+                        "diagnosis_id": CONCEPT.DIAGNOSIS.TARGET_SERVICE_ID,
+                        "external_id": CONCEPT.BIOSPECIMEN_DIAGNOSIS.ID,
+                        "kf_id": CONCEPT.BIOSPECIMEN_DIAGNOSIS.TARGET_SERVICE_ID,
+                        "visible": CONCEPT.BIOSPECIMEN_DIAGNOSIS.VISIBLE,
+                    }
+                )
+                study_merged_df = outer_merge(
+                    study_merged_df,
+                    biospecimen_diagnoses,
+                    with_merge_detail_dfs=False,
+                    on=CONCEPT.DIAGNOSIS.TARGET_SERVICE_ID,
+                )
+
+            # biospecimens
+            biospecimens = study_mapped_df_dict["biospecimens"]
+            biospecimens = biospecimens.rename(
+                columns={
+                    "participant_id": CONCEPT.PARTICIPANT.TARGET_SERVICE_ID,
+                    "sequencing_center_id": CONCEPT.SEQUENCING.CENTER.TARGET_SERVICE_ID,
+                    "age_at_event_days": CONCEPT.BIOSPECIMEN.EVENT_AGE_DAYS,
+                    "analyte_type": CONCEPT.BIOSPECIMEN.ANALYTE,
+                    "composition": CONCEPT.BIOSPECIMEN.COMPOSITION,
+                    "consent_type": CONCEPT.BIOSPECIMEN.CONSENT_SHORT_NAME,
+                    "dbgap_consent_code": CONCEPT.BIOSPECIMEN.DBGAP_STYLE_CONSENT_CODE,
+                    "external_aliquot_id": CONCEPT.BIOSPECIMEN.ID,
+                    "external_sample_id": CONCEPT.BIOSPECIMEN_GROUP.ID,
+                    "kf_id": CONCEPT.BIOSPECIMEN.TARGET_SERVICE_ID,
+                    "method_of_smaple_procurement": CONCEPT.BIOSPECIMEN.SAMPLE_PROCUREMENT,
+                    "ncit_id_anatomical_site": CONCEPT.BIOSPECIMEN.NCIT_ANATOMY_SITE_ID,
+                    "ncit_id_tissue_type": CONCEPT.BIOSPECIMEN.NCIT_TISSUE_TYPE_ID,
+                    "source_text_anatomical_site": CONCEPT.BIOSPECIMEN.ANATOMY_SITE,
+                    "source_text_tissue_type": CONCEPT.BIOSPECIMEN.TISSUE_TYPE,
+                    "source_text_tumor_descriptor": CONCEPT.BIOSPECIMEN.TUMOR_DESCRIPTOR,
+                    "spatial_descriptor": CONCEPT.BIOSPECIMEN.SPATIAL_DESCRIPTOR,
+                    "uberon_id_anatomical_site": CONCEPT.BIOSPECIMEN.UBERON_ANATOMY_SITE_ID,
+                    "visible": CONCEPT.BIOSPECIMEN.VISIBLE,
+                    "volume_ul": CONCEPT.BIOSPECIMEN.VOLUME_UL,
+                }
+            )
+            on = [CONCEPT.PARTICIPANT.TARGET_SERVICE_ID]
+            study_all_targets.update([SequencingCenter, Specimen])
+
+            if biospecimen_diagnoses is not None:
```

Comment:
Wondering why you need this block?

Reply:
As commented above, `/biospecimen-diagnoses` is a conditional endpoint (it exists for cancer studies where tumor specimens need to be associated with diagnoses). Therefore, the above lines first set the participant KF ID to the join-on key because most of the studies (_i.e._, non-cancer studies) will need to join the previously created outer-merged data frame and `biospecimens` only on the participant KF ID. Then, Line 374 checks if `biospecimen_diagnoses` exists and append the biospecimen KF ID to the join-on key list. This is because when `biospecimen_diagnoses` exists, we need to join the previously created outer-merged data frame and `biospecimens` on the participant KF ID AND biospecimen KF ID.